### DISCUSSION TITLE
Will changes to OpenWeather API affect weather info coming into SageTV?

### DISCUSSION THREAD

**Discussion by way-lo** [OpenSageTV/Community#129] [Q&A] [answered]
I believe many of us utilize OpenWeather for our SageTV weather information.  As I recall, we needed to obtain a user-specific passcode and write it into a text file stored in the SageTV directory.  This allows us this to access the weather.

I recently received a notice they are transitioning from One Call 2.5 to 3.0.  The notice is summarized here:
https://openweathermap.org/one-call-transfer

Will this affect if we are able to continue with weather info?

**Comment by jusjoken** [comment]
I have not reviewed the code for a while but at a glance this looks like a
breaking change if they disable the 2.5 version of the api.  My coding and
test servers are still in boxes from our move so not sure if/when I can
take this on.

Are there enough users interested in keeping the weather working?

k

**Reply by way-lo** [reply]
Hi @jusjoken,

Thanks for commenting and taking a look at this.

I would make a gentle request for it to be looked at eventually, time permitting.  Especially now with the info that your test environment is temporarily dismantled.

My personal opinion is the weather forecast is a "nice to have."  The weather icon is a notable and welcome part of the UI, and if it breaks, having an inactive icon mars the appearance.  (I suppose we could just deactivate the plugin when this happens).

**Reply by MechlingBurgh** [reply]
I like the weather but is diffidently lower on my list now a days. I agree if you have limited time I think other things would take priority.

**Reply by mickgeller** [reply]
You can fix the Baseball issue. The root cause is the assigned Category, which is now Baseball only.  Configure it at  Detailed Setup | Customize | TV Category Background Colors. Add Baseball as a category, and assign it your favorite color.

**Reply by Bionicmerlin** [reply]
I like the weather app, but if it comes down to priority, I'd rate anything that brings in video content as more urgent.

**Reply by mickgeller** [reply]
After V 2.5 is disabled, is there a way to prevent function calls, and possibly p*ssing off the API owner?

**Reply by LoisGNS** [reply]
I gave up on the weather function years ago. I remember something that was supposed to fix it, but I couldn't get that working. Lots of other ways to check the weather, so I stopped trying to make it work.

**Comment by jusjoken** [comment]
I believe if you do not have an api key stored then that will prevent
weather calls.

k

**Reply by way-lo** [reply]
@jusjoken Certainly it would be nice not to bombard them with useless traffic if SageTV won't be able to utilize the data. Do you recall how often the weather calls are made in SageTV?

I just looked over an update email from OpenWeather.  It does say all current API keys will continue to work, but only after subscribing to v3, which allows 1000 calls/day for free.

**Comment by jusjoken** [comment]
If I recall its every 5 minutes and occasionally in between depending on
other events or circumstances.  Well within the free 1000 per day for a
users own api key?

In reality this should be a minimal fix once i have my coding environment
up so my plan will be to fix it once i can and push an update which should
fix both gemstone and the core users.

No promise as to when but I might as well fix it.

k

**Reply by emveepee** [reply]
I don't know how SageTV plugins work but I checked the code in the jar file with fernflower and can't the class just be recompiled with the v3 parameter and the jar file updated manually by those that would like it?

**Comment by jusjoken** [comment]
Likely that's the only change needed.

**Comment by jusjoken** [comment]
The code for this is on github in the following location...

https://github.com/jusjoken/GWeather/blob/ba3e318179831319d1c6a42ec96703e5702f7f6f/src/main/sage/sageweather/OWMWeatherLocation.java#L133

If anyone wants to make the change and build it before I get around to it you could clone the repo and make the change and then update the plugin.  I likely can update the plugin itself if someone builds the jar with the change.

k

**Comment by Bionicmerlin** [comment]
Mine quit working after I installed the latest Windows Update yesterday.  It displays 0 degrees.

**Reply by way-lo** [reply]
That used to happen intermittently even when the service was still up.  My server still shows the weather but a Windows client does not.

I last received an email from the service on 10/7:

One Call API 2.5 to be discontinued soon 

Dear OpenWeather client,

We would like to inform you that we started One Call 2.5 deprecation, and your access to the product will be stopped on October 14, 2024. 

We would like to recommend you to switch to One Call 3.0 that would provide you with advanced features such as a hyperlocal minute forecast for one hour, 2-day hourly forecast, 8-day daily forecast, 4-day forecast for any timestamp, government weather alerts, 45-year history for any timestamp, daily aggregation, weather overview, and other features. 

One Call 3.0 has a flexible pay-as-you-call subscription with 1000 free calls per day. Your payment card will be used when you exceed the free limit, £0.0012 per call. API call limits are set up in your account Profile section.

**Comment by ardyssey** [comment]
I built the jar with the One Call 3.0 URL and it works on my SageTV. I also changed the forecast Hi and Lows to use 'max' and 'min' instead of 'day' and 'night'. That works too. 

I signed up for One Call 3.0. In the account you can set your daily call limit. I set mine to 1000. I didn't have to change the API key; the original one still works.

**Comment by jusjoken** [comment]
If you can get me the jar I can update the plugin so others can use this
update.

k

**Reply by ardyssey** [reply]
I uploaded the jars I built to ardyssey/GWeatherFixes

I guess the formal process would be to fork your repository and provide source changes...

**Comment by jusjoken** [comment]
I have create the changes needed to update the plugin (GoogleWeather and
gemstone).  However, the plugin repo depends on Travis CI to perform the
build of the changes into the repo and our free credits for Travis CI
expired in Dec 2023.  I have sent a support request for additional free
credits so I can get the process running again so I will let everyone know
what that response is when I get it.

k

**Comment by jusjoken** [comment]
It took a bit as the plugin repo would not perform its automatic rebuild using Travis-CI.  They responded with some free credits but it was then failing due to what I believe is an expired and/or deleted Person Access Token from when Sean S built this process MANY years ago.  I created a new PAT once I figured out the issue and I was able to get the plugin repo to rebuild properly.

So in theory the weather update provided by @ardyssey is published for both the default STV and gemstone (I have not tested it yet myself).

k

---
@jusjoken
If you wish to see what I am up to and support my efforts visit my [Patreon page](https://www.patreon.com/user?u=58035567)

**Reply by ardyssey** [reply]
Thanks. I will try it out the new published update soon.

**Comment by ardyssey** [comment]
Yes the OpenWeather API change does break weather info. The plugin repo has been rebuilt with a fix.

**Reply by mickgeller** [reply]
I don't do Java, but the plugin repository ZIP file provides a 305 KB text file that looks like XML. Aren't JAR files supposed to look compiled?

Also, Can someone confirm that this requires the user to go to Openweathermap.org and give them a credit card, in the unlikely event API Calls are > 1000 per month? I guess I am asking for confirmation that the "default" account does not provide V3.0 API access.

I gave them a credit card, courtesy of Privacy.com, to control access in an unlikely event. I also went to ardyssey/GWeatherFixes to get a JAR file. I am now waiting for my API key to be activated. Apparently, API access is not instantaneous upon agreeing to the V3.0 plan.

**Reply by mickgeller** [reply]
I can confirm it works. My API key just now started working. I am using the JAR from ardyssey/GWeatherFixes.  I will switch to the correct one when it's published.

**Reply by jusjoken** [reply]
It's fixed (I hope)

**Reply by ardyssey** [reply]
Open Weather does require a credit card for Open Call 3.0 even for free 1000 calls per day (not per month). You can check your usage at https://home.openweathermap.org/statistics/onecall_30 if you're logged in. For the last 4 days, my test jar was making 6-8 calls per day.

**Reply by way-lo** [reply]
I believe you can also set a max calls/day.  So if it is set to 1000, it will never exceed that to start the billing process.

I also utilized a "virtual card" # that can be limited to the $ amount and turned off at whim.

**Reply by ardyssey** [reply]
You can set your max call on page https://home.openweathermap.org/subscriptions

**Comment by jusjoken** [comment]
I will look into this in the morning...perhaps I grabbed the jar file wrong from github.  I will add it to a new version and update the repo tomorrow morning.

k

**Comment by jusjoken** [comment]
Actually I just dug in and fixed it already (I hope).  New version on the plugin repo now.

k

**Comment by way-lo** [comment]
Thanks everyone for your work!

I see this when I peruse the plugins:
![image](https://github.com/user-attachments/assets/e36385e3-3c80-4bcf-a57a-308115fd0808)

I try updating but am hitting this error:
![image](https://github.com/user-attachments/assets/0bc28574-a4a8-4417-9bcc-9a7a61065b14)

If I download the file @ that address, it appeears the zip is invalid when clicked within Windows 10:
![image](https://github.com/user-attachments/assets/58824e1b-5931-498c-bb82-795ccfe18a28)

I believe the latest .zip 3.1.1 file is corrupted.  (Oddly, I can open the zip using 7-zip separately.  Simply dropping the new file into the JARS folder does not work however).

**Comment by jusjoken** [comment]
Odd.  I rebuilt the zip file and published 3.1.1 of GoogleWeather plugin again.  I spun up a test SageTV server and tested and I was able to update from the old version to 3.1.1

Hope it works for everyone else now.

k

**Reply by way-lo** [reply]
Thank you!  SageTV was able to install the new .zip and after a reboot, weather is restored.  For general info, after subscribing to v3.0, I did not need to update my API passcode from the one used in v2.5.

**Comment by ardyssey** [comment]
I ran the Gemstone and Google Weather plugins on my 64 bit client and seems to be working fine. Thanks K for the builds.

**Reply by ardyssey** [reply]
... meant to say the plugin updates...

**Comment by jusjoken** [comment]
I would like to get the source code on Github updated to include these
changes.  I looks like changing the api is just changing the 2.5 to 3.0
(tested).

Do you mind sharing any other files you changed so I can get the source
current (to enable future changes if any)

Thanks
k

**Reply by ardyssey** [reply]
Yes, I can have something for you next week. I'm planning to fork your GWeather repository and add my changed files to the fork. I won't get to that until at least Sunday. Also Git is a new tool for me.

**Comment by jusjoken** [comment]
Great.  I just revamped the project to use maven instead of the very old
grade version so I will commit that likely tomorrow.

k

**Reply by jusjoken** [reply]
I committed my updates to GitHub.  It includes using the 3.0 api but not the min, max change as I have not looked into what that was specifically.  Go ahead and fork it and if you need assistance let me know at usjusjoken at Gmail .com

**Comment by TGsell1** [comment]
I just signed up for OpenWeather and updated the plugins and it works!   Thanks for all the great work. Sent from my iPhoneOn Nov 1, 2024, at 3:05 PM, Ken Birch ***@***.***> wrote:﻿
Great.  I just revamped the project to use maven instead of the very old

grade version so I will commit that likely tomorrow.



k

**Comment by hugo-lyppens** [comment]
I had trouble getting Weather 3.1.1 to work even with owmkey.properties file in the right location. After troubleshooting for some time I found that the commons-lang3 (version 3.4 as of this writing) library is needed but wasn't installed even when removing and reinstalling Google Weather 3.1.1. I think it must be a missing dependency. Hope it helps someone and I hope it will be added as a required dependency. After installing this library, it is necessary to exit & restart SageTV or SageTVClient UI before it takes. Thanks.

**Reply by jusjoken** [reply]
I just reviewed the code and it uses commons-lang ver 2.4 (which the plugin has a dependency included in the manifest) but has no dependency in the code for commons-lang3.  I don't have time right now to test this out but when I do I will test a fresh install to see what's up.

k

**Reply by way-lo** [reply]
This may or may not be helpful, but I found the same thing with my SageTV Win10 client install.  The weather plugin did not work on its own.  Based on the comments here, I tried installing Gemstone (and whatever dependencies that come along with it) and it started working.  It does looks like commons-lang3 was installed along with.

Oddly, using info from the "remove plugin and dependencies" plugin, commons-lang3 does not list Google Weather as a dependent plugin, but the older "commons-lang" v2.4.1 does.

**Reply by jusjoken** [reply, marked as answer]
Fixed with version 3.1.2.  The 3.1.1 version was built by another user (@ardyssey ) who was great to help out and get a version pushed out as I had not had time, however I believe it included an unintended dependency.  3.1.2 was build directly from the source with the changes @ardyssey implemented for OWM 3.0 and min/max for low/high temps.  I tested on a fresh install and no additional dependency should now be needed.

Let me know if any issues occur.

k

**Reply by way-lo** [reply]
Working great now without commons-lang3, thank you @jusjoken

**Reply by Just-messing-around** [reply]
Just a question on this update.

I've had Open Weather working just fine for a long time. A couple of months ago it just stopped working (I get 0 degrees F with no location).

For this new fix (which I've just installed), so I also need to sign up for the free Open Weather account? Or can I use the previous one (where no CC was needed.....just sign up and get an API key)?

**Reply by hugo-lyppens** [reply]
The same API key from before keeps working. For a credit card just enter perhaps a gift card or virtual card to limit potential charges. First 1,000 calls are free, which is of course more than enough for SageTV weather calls. I did not update my owmkey.properties after signing up for weather.

The 3.1.2 version of the weather plugin works fine now on all my laptops and PCs. Thanks for fixing it.

**Comment by KryptoNyte-Sage** [comment]
JMA, I have not completed the latest setup, but when investigating I was not able to find a way to get an API key on OpenWeather without giving them a CC number, even if my intention was to never have to use it (making over 1000 requests to OpenWeather per day).

I believe you need to follow this link from an earlier post:
https://home.openweathermap.org/statistics/onecall_30
which for me (when logged in) shows that I "do not have an active subscription to OneCall 3.0 API."

As I understand it, we then need to go here to subscribe to OneCall API 3.0:
https://home.openweathermap.org/subscriptions

EDIT: I went ahead and subscribed to the Base Plan with a CC.  Although there is no current invoice or indication of billing soon after, when I went here:
https://home.openweathermap.org/api_keys
it showed me that the API key was "Active."  It is also the same API key that I have been using previously.  After restarting the server, weather still hasn't updated just 15 minutes after subscribing.  I assume I need to give their system some time to catch up.

EDIT2:  After an hour or so, the OneCall subscription was active and I received an email confirming.  I logged back into my account (or use the subscription link above after logging in) I was able to see the active sub and edit/reduce the call limit to less than 1000.  The was also described in the confirmation email.

Thank you, Ken, for getting this working once again.

**Reply by Just-messing-around** [reply]
KryptoNyte-Sage, thank you for the posting and links.....

Thanks to all for this update. 

My solution was simple, with the introduction of the 3.0 API, I simply needed to have a card on file for the free account. I did update the maximum API calls to 900 so that I would never incur a cost.

**Comment by way-lo** [comment]
As a general FYI, you may check your call usage at this link:
https://home.openweathermap.org/statistics/onecall_30

I have 1 server that is used infrequently and 1 client that is used even less frequently.

I am averaging 8-11 calls/day.

The old API that was used for v2 works fine.  I also made a new one for my client just for fun.  Apparently you may generate as many as you want and enable/revoke them at will.
